hudi-cli/src/main/java/org/apache/hudi/cli/commands/DiffCommand.java

```diff
@@ -0,0 +1,196 @@
+  }
+
+  private HoodieDefaultTimeline getTimelineInRange(String startTs, String endTs, boolean includeArchivedTimeline) {
+    if (StringUtils.isNullOrEmpty(startTs)) {
+      startTs = getTimeDaysAgo(10);
+    }
+    if (StringUtils.isNullOrEmpty(endTs)) {
+      endTs = getTimeDaysAgo(1);
+    }
+    HoodieTableMetaClient metaClient = HoodieCLI.getTableMetaClient();
+    HoodieActiveTimeline activeTimeline = metaClient.getActiveTimeline();
+    if (includeArchivedTimeline) {
+      HoodieArchivedTimeline archivedTimeline = metaClient.getArchivedTimeline();
+      archivedTimeline.loadInstantDetailsInMemory(startTs, endTs);
+      return archivedTimeline.findInstantsInRange(startTs, endTs).mergeTimeline(activeTimeline);
+    }
+    return activeTimeline;
+  }
+
+  private String printCommitsWithMetadataForFileId(HoodieDefaultTimeline timeline,
+                                                   final Integer limit,
+                                                   final String sortByField,
+                                                   final boolean descending,
+                                                   final boolean headerOnly,
+                                                   final String tempTableName,
+                                                   final String fileId) throws IOException {
+    final List<Comparable[]> rows = new ArrayList<>();
+
+    final List<HoodieInstant> commits = timeline.getCommitsTimeline().filterCompletedInstants()
+        .getInstants().sorted(HoodieInstant.COMPARATOR.reversed()).collect(Collectors.toList());
+
+    for (final HoodieInstant commit : commits) {
+      if (timeline.getInstantDetails(commit).isPresent()) {
+        final HoodieCommitMetadata commitMetadata = HoodieCommitMetadata.fromBytes(
+            timeline.getInstantDetails(commit).get(),
+            HoodieCommitMetadata.class);
+
+        for (Map.Entry<String, List<HoodieWriteStat>> partitionWriteStat :
+            commitMetadata.getPartitionToWriteStats().entrySet()) {
+          for (HoodieWriteStat hoodieWriteStat : partitionWriteStat.getValue()) {
+            if (fileId.equals(hoodieWriteStat.getFileId())) {
+              rows.add(new Comparable[] {commit.getAction(), commit.getTimestamp(), hoodieWriteStat.getPartitionPath(),
+                  hoodieWriteStat.getFileId(), hoodieWriteStat.getPrevCommit(), hoodieWriteStat.getNumWrites(),
+                  hoodieWriteStat.getNumInserts(), hoodieWriteStat.getNumDeletes(),
+                  hoodieWriteStat.getNumUpdateWrites(), hoodieWriteStat.getTotalWriteErrors(),
+                  hoodieWriteStat.getTotalLogBlocks(), hoodieWriteStat.getTotalCorruptLogBlock(),
+                  hoodieWriteStat.getTotalRollbackBlocks(), hoodieWriteStat.getTotalLogRecords(),
+                  hoodieWriteStat.getTotalUpdatedRecordsCompacted(), hoodieWriteStat.getTotalWriteBytes()
```

Comment:
to reduce duplicate code, this can be encapsulated into HoodieWriteStat to return the Comparable[] ? (less the first 2)

Reply:
i think it's more readable inline. To reduce duplicate code, i can extract to a separate method.